homeassistant/components/icloud/__init__.py

```diff
@@ -1 +1,607 @@
+
+    def __init__(
+        self,
+        hass: HomeAssistantType,
+        username: str,
+        password: str,
+        account_name: str,
+        max_interval: int,
+        gps_accuracy_threshold: int,
+    ):
+        """Initialize an iCloud account."""
+        self.hass = hass
+        self._username = username
+        self._password = password
+        self._name = account_name or slugify(username.partition("@")[0])
+        self._fetch_interval = max_interval
+        self._max_interval = max_interval
+        self._gps_accuracy_threshold = gps_accuracy_threshold
+
+        self.api = None
+        self._owner_fullname = None
+        self._family_members_fullname = {}
+        self._devices = {}
+
+        self.unsub_device_tracker = None
+
+    def reset_account(self):
+        """Reset an iCloud account."""
+        icloud_dir = self.hass.config.path("icloud")
+
+        try:
+            self.api = PyiCloudService(self._username, self._password, icloud_dir)
+        except PyiCloudFailedLoginException as error:
+            self.api = None
+            _LOGGER.error("Error logging into iCloud Service: %s", error)
+            return
+
+        user_info = None
+        try:
+            # Gets device owners infos
+            user_info = self.api.devices.response["userInfo"]
+        except PyiCloudNoDevicesException:
+            _LOGGER.error("No iCloud Devices found")
+
+        self._owner_fullname = f"{user_info['firstName']} {user_info['lastName']}"
+
+        self._family_members_fullname = {}
+        for prs_id, member in user_info["membersInfo"].items():
+            self._family_members_fullname[
+                prs_id
+            ] = f"{member['firstName']} {member['lastName']}"
+
+        self._devices = {}
+        self.update_devices()
+
+    def update_devices(self) -> None:
+        """Update iCloud devices."""
+        if self.api is None:
+            return
+
+        api_devices = {}
+        try:
+            api_devices = self.api.devices
+        except PyiCloudNoDevicesException:
+            _LOGGER.error("No iCloud Devices found")
+
+        # Gets devices infos
+        for device in api_devices:
+            status = device.status(DEVICE_STATUS_SET)
+            device_id = status[DEVICE_ID]
+            device_name = status[DEVICE_NAME]
+
+            if self._devices.get(device_id, None) is not None:
+                # Seen device -> updating
+                _LOGGER.debug("Updating iCloud device: %s", device_name)
+                self._devices[device_id].update(status)
+            else:
+                # New device, should be unique
+                _LOGGER.debug(
+                    "Adding iCloud device: %s [model: %s]",
+                    device_name,
+                    status[DEVICE_RAW_DEVICE_MODEL],
+                )
+                self._devices[device_id] = IcloudDevice(self, device, status)
+                self._devices[device_id].update(status)
```

Comment:
You need to call `update` even after passing `status` to the constructor ?

Reply:
I made it like that following https://github.com/home-assistant/home-assistant/pull/24053#discussion_r301759288 and https://github.com/home-assistant/home-assistant/pull/26590#discussion_r324058381 reviews.

It is to setup the `IcloudDevice`, but maybe I'm wrong.

How can I make it better ?

Reply:
Anything on this @balloob ?

Reply:
Oh it's doing I/O ? In that case it's fine.

Reply:
Yep